src/sonic-supervisord-utilities-rs/src/supervisor_proc_exit_listener.rs

```diff
@@ -0,0 +1,516 @@
+        process::exit(3);
+    }
+
+    let feature_config = feature_config.unwrap();
+    let is_auto_restart = feature_config.get("auto_restart");
+    if is_auto_restart.is_none() {
+        error!("Unable to determine auto-restart feature status for '{}'. Exiting...", container_name);
+        process::exit(4);
+    }
+
+    Ok(is_auto_restart.unwrap().clone())
+}
+
+/// Load heartbeat alert intervals from ConfigDB
+pub fn load_heartbeat_alert_interval(config_db: &dyn ConfigDBTrait) -> Result<()> {
+
+    let heartbeat_table = config_db.get_table(HEARTBEAT_TABLE_NAME)
+        .map_err(|e| SupervisorError::Database(format!("Failed to get HEARTBEAT table: {}", e)))?;
+
+    if !heartbeat_table.is_empty() {
+        let mapping = HEARTBEAT_ALERT_INTERVAL_MAPPING.get_or_init(|| Mutex::new(HashMap::new()));
+        let mut mapping = mapping.lock().unwrap();
+        
+        for (process, config) in heartbeat_table {
+            if let Some(alert_interval_str) = config.get("alert_interval") {
+                if let Ok(alert_interval_ms) = alert_interval_str.parse::<i64>() {
+                    // Convert from milliseconds to seconds
+                    mapping.insert(process, alert_interval_ms as f64 / 1000.0);
+                }
+            }
+        }
+    }
+
+    HEARTBEAT_ALERT_INTERVAL_INITIALIZED.set(true).map_err(|_| {
+        SupervisorError::System("Failed to set heartbeat initialized flag".to_string())
+    })?;
+
+    Ok(())
+}
+
+/// Get heartbeat alert interval for process
+pub fn get_heartbeat_alert_interval(process: &str, config_db: &dyn ConfigDBTrait) -> f64 {
+    if HEARTBEAT_ALERT_INTERVAL_INITIALIZED.get().is_none() {
+        if let Err(e) = load_heartbeat_alert_interval(config_db) {
+            warn!("Failed to load heartbeat alert intervals: {}", e);
+        }
+    }
+
+    let mapping = HEARTBEAT_ALERT_INTERVAL_MAPPING.get_or_init(|| Mutex::new(HashMap::new()));
+    if let Ok(mapping) = mapping.lock() {
+        if let Some(&interval) = mapping.get(process) {
+            return interval;
+        }
+    }
+
+    ALERTING_INTERVAL_SECS as f64
+}
+
+/// Publish events
+pub fn publish_events(events_handle: &EventPublisher, process_name: &str, container_name: &str) -> Result<()> {
+    let mut params = HashMap::new();
+    params.insert("process_name".to_string(), process_name.to_string());
+    params.insert("ctr_name".to_string(), container_name.to_string());
+    
+    events_handle.publish(EVENTS_PUBLISHER_TAG, Some(&params))
+        .map_err(|e| SupervisorError::System(format!("Failed to publish event: {}", e)))?;
+    
+    info!("Published event: {} for process {} in container {}", EVENTS_PUBLISHER_TAG, process_name, container_name);
+    Ok(())
+}
+
+/// Get current time as Unix timestamp - helper function
+pub fn get_current_time() -> f64 {
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap_or_default()
+        .as_secs_f64()
```

Comment:
This is going to be vulnerable to time jumps. I recommend using a monotonic time if possible.